Add 3D support to feature detection and tracking

freemansw1 opened the pull request:
This is split off from #127. Adds 3D support to feature detection and tracking, and adds support for these new 3D features in segmentation. 

Question for @JuliaKukulies -- I've disabled wavelength filtering for input 3D data. Does this make sense? If not, what would be the best way to incorporate this filtering into 3D data? Should it be applied to all levels, just to some levels, etc.?

This was done collaboratively with @galexsky

* [x] Have you followed our guidelines in CONTRIBUTING.md? 
* [x] Have you self-reviewed your code and corrected any misspellings? 
* [x] Have you written documentation that is easy to understand?
* [x] Have you written descriptive commit messages? 
* [x] Have you added NumPy docstrings for newly added functions? 
* [x] Have you formatted your code using black? 
* [x] If you have introduced a new functionality, have you added adequate unit tests?
* [x] Have all tests passed in your local clone? 
* [ ] If you have introduced a new functionality, have you added an example notebook?
* [x] Have you kept your pull request small and limited so that it is easy to review? 
* [x] Have the newest changes from this branch been merged?

snilsn reviewed (changes requested):
This took a while, but I'm through now @freemansw1. Great extension to tobac, and I think overall fully functional. Most of the things I found are only cosmetic in nature, e. g. to actually use implemented utilities to make the code clearer.

Some bugs need to be adressed though, e. g. my first comment about `feature_detection_multithreshold`.

JuliaKukulies commented on tobac/tests/test_segmentation.py:
```diff
@@ -112,3 +169,198 @@ def test_segmentation_timestep_level():
+        )
+    elif vertical_axis_num == 2:
+        test_data_iris = testing.make_dataset_from_arr(
+            z_dim_num=1, y_dim_num=0, x_dim_num=1, **common_dset_opts
```
`zdim_num = 2` ; if this is changed to the correct axis number the test fails, meaning that the bug related to the transposing in `segmentation_timestep()`is successfully captured

freemansw1 replied:
I've fixed this, it did catch the bug, and I've now fixed the bug.

JuliaKukulies reviewed (changes requested):
Such a great contribution, very well done @freemansw1 and @galexsky! I am sorry this took longer than expected but I finished going through this PR now and really appreciated that you separated this from the PBC changes. This time it was much easier to systematically test the main features related to 3D feature detection, segmentation, and tracking.  

I found a couple of small bugs testing these new features with my own data. Once they are solved we should probably also make sure that they are covered by tests. Notice also that for some reason the API docs are not correctly rendered in this PR. 

As discussed and decided in the last developer meeting, disabling the wavelength filtering is reasonable for 3D input data.

freemansw1 commented:
Thanks @JuliaKukulies and @snilsn for your detailed reviews! Now that I'm back from AMS, I'll work to resolve these later this week.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -598,24 +725,79 @@ def feature_detection_multithreshold(
+            # is a lower axis number than the specified vertical coordinate.
+            if ndim_time < vertical_axis:
+                vertical_axis = vertical_axis - 1
+        else:
```
Ah, this documentation was conflicting. I've clarified now that `vertical_coord` should be a string.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -598,24 +725,79 @@ def feature_detection_multithreshold(
+                vertical_axis = vertical_axis - 1
+        else:
+            # We need to determine vertical axis
+            vertical_axis = internal_utils.find_vertical_axis_from_coord(
```
Good catch. This is now resolved.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -598,24 +725,79 @@ def feature_detection_multithreshold(
+        Name or axis number of the vertical coordinate. If 'auto', tries to auto-detect.
+        It looks for the coordinate or the dimension name corresponding
+        to the string.
+    vertical_axis: int or None.
```
I think I have resolved this by just checking if both are set, and if so, raising a `ValueError`. I've also added a test to this effect.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -283,92 +319,157 @@ def feature_detection_threshold(
-            # on tuples, let's convert to 1D coordinates.
-            region_i = np.array(hdim1_indices * x_size + hdim2_indices)
+            if is_3D:
+                region_i = list(
```
Good catch. I've moved both to list for consistency.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -283,92 +319,157 @@ def feature_detection_threshold(
+            if is_3D:
+                region_i = list(
+                    zip(
+                        hdim1_indices * (x_max + 1) * (z_max + 1)
```
I've included a comment block above this to help out here.

freemansw1 commented on tobac/feature_detection.py:
````diff
@@ -693,50 +884,141 @@ def feature_detection_multithreshold(
+        even if ```x_coordinate_name``` and ```y_coordinate_name``` are specified.
+    dz: float
+        Constant vertical grid spacing (m), optional. If not specified
+        and the input is 3D, this function requires that `altitude` is available
````
Good catch.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -693,50 +884,141 @@ def feature_detection_multithreshold(
+    features:      pandas DataFrame
+                   features
+    dxy:           float
+        Constant horzontal grid spacing (m), optional. If not specified,
```
Good catch, resolved.

freemansw1 commented on tobac/utils/internal.py:
```diff
@@ -407,3 +413,225 @@ def wrapper(*args, **kwargs):
+        coords_1 = np.array((0, coords_1[0], coords_1[1]))
+        coords_2 = np.array((0, coords_2[0], coords_2[1]))
+
+    deltas = np.abs(coords_1 - coords_2)
```
Good catch; squaring things in the next line removes the need for `np.abs`. Resolved.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -693,50 +884,141 @@ def feature_detection_multithreshold(
+                    dxy * features.loc[index_2, "hdim_2"],
+                )
+
+            distance = internal_utils.calc_distance_coords(
```
This is a good point! I am going to raise a separate issue on this; the current code behaves this way, too, so we should address it there. I'm disinclined from changing this behavior with the addition of 3D.

freemansw1 replied:
New issue #230 addresses this. I'm going to mark this conversation as resolved, and we can deal with it over there.

freemansw1 commented on tobac/tracking.py:
```diff
@@ -152,6 +160,12 @@ def linking_trackpy(
         Default is -1
 
+    vertical_coord: str
+        Name of the vertical coordinate in meters. If 'auto', tries to auto-detect.
```
Yeah, that language was ambiguous. I've updated it and resolved here.

freemansw1 commented on tobac/tracking.py:
```diff
@@ -242,20 +293,32 @@ def linking_trackpy(
+            dist_func=dist_func,
         )
     elif method_linking == "predict":
+        if is_3D and pkgvsn.parse(tp.__version__) < pkgvsn.parse("0.6.0"):
```
I agree with your sentiment here, but I also don't want us to have to go back and edit while we wait on trackpy. I'm happy to just throw an error, but we will have to be careful to release a new version as soon as trackpy releases.

snilsn replied:
Ok, considering how small this issue is, I think I would be happy to leave it as is and hope for a `trackpy` release soon.

freemansw1 commented on tobac/tracking.py:
```diff
@@ -304,10 +384,7 @@ def linking_trackpy(
-    remap_particle_to_cell_vec = np.vectorize(
-        lambda particle_cell_map, input_particle: particle_cell_map[input_particle]
-    )
+    remap_particle_to_cell_vec = np.vectorize(remap_particle_to_cell_nv)
```
Yeah, the particle numbers are not necessarily contiguous, so this doesn't work if we want contiguous cell numbers. My guess is that there's an easier solution here, but I don't think this code will work to provide contiguous cell numbers. 

It's worth noting that we don't guarantee contiguous cell numbers; in fact, cell numbers won't be contiguous in many cases, but that's a separate issue. This will at least result in cell numbers starting at ~cell_number_start and ~incrementing from there (the particle numbers can end up quite large).

snilsn replied:
Ok I see, I did not think about this. Then we should go with this working solution as it is.

freemansw1 commented on tobac/tracking.py:
```diff
@@ -461,3 +540,18 @@ def add_cell_time(t):
     return t
+
+
+def remap_particle_to_cell_nv(particle_cell_map, input_particle):
```
See above comment

freemansw1 commented on tobac/utils/general.py:
```diff
@@ -147,26 +148,190 @@ def add_coordinates(t, variable_cube):
+def add_coordinates_3D(
+    t, variable_cube, vertical_coord="auto", assume_coords_fixed_in_time=True
+):
+    import numpy as np
```
Good point. I've moved all of the numpy imports to the top of the file here. We do need to address this module-wide, though (#188)

freemansw1 commented on tobac/utils/general.py:
```diff
@@ -147,26 +148,190 @@ def add_coordinates(t, variable_cube):
+    # chose right dimension for horizontal and vertical axes based on time dimension:
+    ndim_time = variable_cube.coord_dims("time")[0]
+
+    # TODO: move this to a function, this is duplicated from segmentation.
```
I decided against this as this specific functionality is probably unique enough to not be a function. I have, however, moved this to use the `internal_utils.find_axis_from_coord` function, which I think makes this read better.

freemansw1 commented on tobac/utils/internal.py:
```diff
@@ -407,3 +413,225 @@ def wrapper(*args, **kwargs):
+    elif len(all_matching_axes) > 1:
+        raise ValueError("Too many axes matched.")
+    else:
+        return -1
```
I've switched to `None` here. You're right that -1 is a valid index, but I don't want to raise an error here, as sometimes no axis found may be a valid choice.

freemansw1 commented on tobac/utils/internal.py:
```diff
@@ -407,3 +413,225 @@ def wrapper(*args, **kwargs):
+        )
+        ndim_vertical = field_in.coord_dims(vertical_axis)
+        if len(ndim_vertical) > 1:
+            raise ValueError("please specify 1 dimensional vertical coordinate")
```
Yes; these are intended as internal use only, so the value error here is primarily for future developers.

freemansw1 commented on tobac/utils/general.py:
```diff
@@ -147,26 +148,190 @@ def add_coordinates(t, variable_cube):
+
+        # interpolate 3D coordinates:
+        elif var_coord.ndim == 3:
+            curr_coord_dims = variable_cube.coord_dims(coord)
```
I believe that this can be triggered? It's not tested, though, which isn't great.

freemansw1 replied:
Removed

freemansw1 commented on tobac/utils/internal.py:
```diff
@@ -407,3 +413,225 @@ def wrapper(*args, **kwargs):
+
+
+def find_hdim_axes_3D_iris(field_in, vertical_coord=None):
+    """Finds what the hdim axes are given a 3D (including z) or
```
This has been fixed

freemansw1 commented on tobac/segmentation.py:
```diff
@@ -168,15 +359,59 @@ def segmentation_timestep(
+        # horizontal coordinates
+        # To make things easier, we will transpose the axes
+        # so that they are consistent.
+        if vertical_coord_axis == 0:
```
Moved to use this utility.

freemansw1 commented on tobac/segmentation.py:
```diff
@@ -32,6 +32,181 @@
+                z_seed_start = int(np.max([0, np.ceil(row["vdim"] - seed_z / 2)]))
+                z_seed_end = int(np.min([z_len, np.ceil(row["vdim"] + seed_z / 2)]))
+
+            hdim_1_min = int(np.ceil(row["hdim_1"] - seed_h1 / 2))
```
I'm not sure what you mean here.

snilsn replied:
It's been a while, but I think I was referring to using `np.min' and `np.max' in line 126/125 to obey the limits of the array. Does this make sense, do we need this for hdim1 and hdim2 min/max as well?

freemansw1 commented on tobac/testing.py:
```diff
@@ -584,6 +601,8 @@ def make_feature_blob(
 
     shape: str('rectangle'), optional
         The shape of the blob that is added. For now, this is just rectangle
+        'oval' adds an oval/spherical bubble with constant amplitude `amplitude`. We assume that the
```
Removed.

freemansw1 commented on tobac/tests/test_feature_detection.py:
```diff
@@ -148,6 +148,248 @@ def test_feature_detection_position(position_threshold):
+    else:
+        # Expecting a raise
+        with pytest.raises(ValueError):
+            out_seg_mask, out_df = feat_detect.feature_detection_multithreshold(
```
Good catch!

freemansw1 commented on doc/feature_detection_output.rst:
```diff
@@ -9,3 +9,10 @@ Variables that are common to all feature detection files:
+Variables that are included when using 3D feature detection in addition to those above:
+
+.. csv-table:: tobac 3D Feature Detection Output Variables
+   :file: ./feature_detection_3D_out_vars.csv
```
Good catch. I've added this.

freemansw1 commented on doc/data_input.rst:
```diff
@@ -1,16 +1,27 @@
+3D Data
+=======
+
+As of *tobac* version 1.5.0, 3D data are now fully supported for feature detection, tracking, and segmentation. Similar to how *tobac* requires some information on the horizontal grid spacing of the data (e.g., through the :code:`dxy` parameter), some information on the vertical grid spacing is also required. This is documented in detail in the API docs, but briefly, users must specify either :co ...
```
I was able to fix this by adding `sklearn` as a mock import to our `conf.py` file. I don't know why that didn't break before, but hopefully this resolves it for good.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -283,92 +319,157 @@ def feature_detection_threshold(
+            )
+        else:
+            selem = np.ones((n_erosion_threshold, n_erosion_threshold))
+        mask = binary_erosion(mask, selem).astype(bool)
```
Good point. I've removed this.

freemansw1 commented on tobac/tests/test_testing.py:
```diff
@@ -127,35 +149,169 @@ def test_generate_single_feature():
+    expected_outs: array-like, either 1D, 2D, or 3D
+        The expected output
+    """
+    import numpy as np
```
Moved.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -598,24 +725,79 @@ def feature_detection_multithreshold(
+            # We only need to adjust the axis number if the time axis
+            # is a lower axis number than the specified vertical coordinate.
+            if ndim_time < vertical_axis:
+                vertical_axis = vertical_axis - 1
```
Good point. I've added this in the docstring and added validation.

freemansw1 commented on tobac/tests/test_utils_internal.py:
```diff
@@ -0,0 +1,62 @@
+import tobac.utils.internal as internal_utils
```
I think I have clarified this enough such that a separate test isn't needed? This function is covered elsewhere, as well. I'm happy to hear pushback on this, though!

JuliaKukulies replied:
You are right, I think it should be clear enough now and I probably missed that the this code is covered through the other tests

freemansw1 commented on tobac/segmentation.py:
```diff
@@ -261,6 +458,16 @@ def segmentation_timestep(
+    segmentation_mask[~unmasked] = -1
+
+    if transposed_data:
+        segmentation_mask = np.transpose(
```
Excellent catch. I've resolved this and fixed the testing issue too.

freemansw1 commented:
Thanks for your very helpful comments, @snilsn and @JuliaKukulies . I've (hopefully) addressed them all, or at least acknowledged them, and I'm now ready for a re-review.

w-k-jones commented on tobac/segmentation.py:
```diff
@@ -32,6 +32,181 @@
+        for index, row in features.iterrows():
+            marker_arr[level, int(row["hdim_1"]), int(row["hdim_2"])] = row["feature"]
+
+    elif seed_3D_flag == "box":
```
Should raise an exception if seed_3D_flag is neither "column" nor "box"

freemansw1 replied:
Good point, I've added this check up front.

w-k-jones commented on tobac/tracking.py:
```diff
@@ -270,22 +341,35 @@ def linking_trackpy(
+            trajectories_unfiltered.rename(
+                columns={"y": "hdim_1", "x": "hdim_2"}, inplace=True
+            )
+            features_linking.rename(
```
features_linking doesn't necessarily need the columns renamed as it shouldn't be returned by the function

freemansw1 replied:
I believe that this is resolved by the merge conflict change

w-k-jones reviewed (commented):
Looks good overall. All of my comments are minor so can be ignored and fixed at a later date

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -86,20 +90,38 @@ def feature_position(
 
     Returns
     -------
+    float
```
Good point, fixed.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -86,20 +90,38 @@ def feature_position(
+            region_bbox[1] : region_bbox[4],
+            region_bbox[2] : region_bbox[5],
+        ]
 
```
Good point.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -125,12 +151,18 @@ def feature_position(
         )
-    return hdim1_index, hdim2_index
+
+    if is_3D:
```
I am hesitant to do that as there will be a lot of duplicated code and docs. We could separate out just the vertical position, but I do think that could cause problems with future position finding functions that may require information about all three coordinates simultaneously.

w-k-jones replied:
Good point, I think leave it as is and it perhaps be something we revisit in the future

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -283,92 +319,164 @@ def feature_detection_threshold(
+
+    # We need to transpose the input data
+    if is_3D:
+        if vertical_axis == 1:
```
Yes, good point.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -283,92 +319,164 @@ def feature_detection_threshold(
+        mask = data_i <= threshold
     # only include values greater than threshold
     # erode selected regions by n pixels
     if n_erosion_threshold > 0:
```
Agreed, and we should probably do that anyway. Can you make a new issue so that we don't forget? We can tag it low priority.

w-k-jones replied:
Will do

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -283,92 +319,164 @@ def feature_detection_threshold(
+                    zip(
+                        hdim1_indices * (x_max + 1) * (z_max + 1)
+                        + hdim2_indices * (z_max + 1)
+                        + vdim_indices
```
Good find; numpy really does have a function for everything. I've replaced this code with the numpy equivalent.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -642,24 +776,97 @@ def feature_detection_multithreshold(
-            FutureWarning,
+    logging.debug("start feature detection based on thresholds")
+
+    if "time" not in [coord.name() for coord in field_in.coords()]:
```
Probably, but I'm hesitant to do so here with the impending switch to xarray.

freemansw1 commented on tobac/testing.py:
```diff
@@ -3,6 +3,7 @@
 """
 
 import datetime
+from multiprocessing.sharedctypes import Value
```
It is not, so I have removed it.

freemansw1 commented on tobac/tracking.py:
```diff
@@ -247,20 +306,32 @@ def linking_trackpy(
         # avoid setting pos_columns by renaimng to default values to avoid trackpy bug
-        features.rename(columns={"hdim_1": "y", "hdim_2": "x"}, inplace=True)
+        if not is_3D:
+            features_linking.rename(
```
I believe that this merge conflict has been fixed.

freemansw1 commented:
@w-k-jones, I think I have addressed all of your comments. Let me know what you think!

In terms of merging, I'm hesitant to do so until at least @JuliaKukulies has a chance to make sure I addressed her comments, but I know she's quite busy. @snilsn you are welcome to continue reviewing/editing, but given your change in jobs, if you want to drop off, I totally understand!

w-k-jones commented:
I’m happy with the changes made, I think any residual issues are both minor and better suited for a future refactor when we convert to xarray internally

snilsn reviewed (approved):
@freemansw1 I think you've addressed every issue I found, and I'm happy with the current state of this update, so I'm approving the PR now. 

However, I did not (and will not) go through it again in detail, so I don't know if this counts as a full approval.

JuliaKukulies commented:
> @w-k-jones, I think I have addressed all of your comments. Let me know what you think!
> 
> In terms of merging, I'm hesitant to do so until at least @JuliaKukulies has a chance to make sure I addressed her comments, but I know she's quite busy. @snilsn you are welcome to continue reviewing/editing, but given your change in jobs, if you want to drop off, I totally understand!

I went through your answers and the commits that concern my requests and I am happy with these changes! Like Nils, I did not do a complete re-review, but maybe this is OK given that @w-k-jones also had a look? If really needed, I will try to do so on Friday but I am OK with merging right away
